pkg/selector/selector.go

```diff
@@ -201,20 +212,128 @@ func (itf Selector) FilterInstanceTypes(filters Filters) ([]*instancetypes.Detai
+			}
+		} else if *sortFilter == sortSpotPrice {
+			if err := itf.EC2Pricing.RefreshSpotCache(spotPricingDaysBack); err != nil {
+				return nil, fmt.Errorf("there was a problem refreshing the spot pricing cache: %v", err)
+			}
+		}
 	}
-	sort.Slice(instanceTypeInfoSlice, func(i, j int) bool {
-		iInstanceInfo := instanceTypeInfoSlice[i]
-		jInstanceInfo := instanceTypeInfoSlice[j]
-		return strings.Compare(aws.StringValue(iInstanceInfo.InstanceType), aws.StringValue(jInstanceInfo.InstanceType)) <= 0
+
+	// sort instance types based on filter flag and isDescending
+	isSortingFlagValid := true
+	sort.Slice(instanceTypes, func(i, j int) bool {
+		firstType := instanceTypes[i]
+		secondType := instanceTypes[j]
+
+		// Determine which value to sort by.
+		// Handle nil values by making non nil values always less than the nil values. That way the
+		// nil values can be bubbled up to the end of the list.
+		switch *sortFilter {
+		case sortODPrice:
+			if firstType.OndemandPricePerHour == nil {
+				return false
+			} else if secondType.OndemandPricePerHour == nil {
+				return true
+			}
+
+			if isDescending {
+				return *firstType.OndemandPricePerHour > *secondType.OndemandPricePerHour
+			} else {
+				return *firstType.OndemandPricePerHour <= *secondType.OndemandPricePerHour
+			}
+		case sortSpotPrice:
+			if firstType.SpotPrice == nil {
+				return false
+			} else if secondType.SpotPrice == nil {
+				return true
+			}
+
+			if isDescending {
+				return *firstType.SpotPrice > *secondType.SpotPrice
+			} else {
+				return *firstType.SpotPrice <= *secondType.SpotPrice
+			}
+		case sortVcpu:
+			if firstType.VCpuInfo == nil || firstType.VCpuInfo.DefaultVCpus == nil {
+				return false
+			} else if secondType.VCpuInfo == nil || secondType.VCpuInfo.DefaultVCpus == nil {
+				return true
+			}
+
+			if isDescending {
+				return *firstType.VCpuInfo.DefaultVCpus > *secondType.VCpuInfo.DefaultVCpus
+			} else {
+				return *firstType.VCpuInfo.DefaultVCpus <= *secondType.VCpuInfo.DefaultVCpus
+			}
+		case sortMemory:
+			if firstType.MemoryInfo == nil || firstType.MemoryInfo.SizeInMiB == nil {
+				return false
+			} else if secondType.MemoryInfo == nil || secondType.MemoryInfo.SizeInMiB == nil {
+				return true
+			}
+
+			if isDescending {
+				return *firstType.MemoryInfo.SizeInMiB > *secondType.MemoryInfo.SizeInMiB
+			} else {
+				return *firstType.MemoryInfo.SizeInMiB <= *secondType.MemoryInfo.SizeInMiB
+			}
+		case sortName:
+			if firstType.InstanceType == nil {
+				return false
+			} else if secondType.InstanceType == nil {
+				return true
+			}
+
+			if isDescending {
+				return strings.Compare(aws.StringValue(secondType.InstanceType), aws.StringValue(firstType.InstanceType)) <= 0
+			} else {
+				return strings.Compare(aws.StringValue(firstType.InstanceType), aws.StringValue(secondType.InstanceType)) <= 0
+			}
+		default:
+			// invalid sorting flag. Do not sort.
+			isSortingFlagValid = false
```

Comment:
It looks like this would keep sorting the entire list of instance types, despite an invalid sorting flag. It would be better to check for this earlier on to skip or short-circuit any unnecessary sorting.

Reply:
That is a very good point. I will ensure the sorting flag is validated before `sort.Slice()` is called so that an early return can occur if needed.